src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp

```diff
@@ -0,0 +1,245 @@
+    "DictionaryEncoding, BagOfWordsEncoding and TfIdfEncoding methods.",
+    // Long description.
+    "This utility takes a dataset and the dimension and arguments and "
+    "encodes string dataset according to arguments given."
+    "\n\n"
+    "The dataset may be given as the file name and the output may be saved as "
+    + PRINT_PARAM_STRING("actual_dataset") + " and " +
+    PRINT_PARAM_STRING("preprocess_dataset") + " ."
+    "\n\n"
+    " Following arguments may be given " + PRINT_PARAM_STRING("encoding_type") +
+    " to encode the dataset using a specific encoding type and " + " Also the "
+    "dimension which contains the string dataset " +
+    PRINT_PARAM_STRING("dimension") + "."
+    "\n\n"
+    "So, a simple example where we want to encode string dataset " +
+    PRINT_DATASET("X") + ", which is having string data in its 3 Column,"
+    " using DictionaryEncoding as encoding type."
+    "\n\n" +
+    PRINT_CALL("preprocess_string", "actual_dataset", "X",
+        "preprocess_dataset", "X", "dimension", 3, "encoding_type",
+        "DictionaryEncoding") +
+    "\n\n",
+    SEE_ALSO("@preprocess_binarize", "#preprocess_binarize"),
+    SEE_ALSO("@preprocess_describe", "#preprocess_describe"),
+    SEE_ALSO("@preprocess_imputer", "#preprocess_imputer"));
+
+PARAM_STRING_IN_REQ("actual_dataset", "File containing the reference dataset.",
+    "t");
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess or"
+    "encoded dataset.", "o");
+PARAM_STRING_IN("column_delimiter", "delimeter used to seperate Column in files"
+    "example '\\t' for '.tsv' and ',' for '.csv'.", "d", "\t");
+PARAM_STRING_IN("delimiter", "A set of chars that is used as delimeter to"
+    "tokenize the string dataset ", "D", " ");
+PARAM_STRING_IN("encoding_type", "Type of encoding", "e", "DictionaryEncoding")
+PARAM_STRING_IN("tfidf_encoding_type", "Type of tfidf encoding", "E",
+    "RawCount")
+PARAM_FLAG("smooth_idf", "True to have smooth_idf for Tf-Idf.", "s");
+PARAM_VECTOR_IN_REQ(std::string, "dimension", "Column which contains the"
+    "string data. (1 by default)", "c");
+
+using namespace mlpack;
+using namespace mlpack::util;
+using namespace arma;
+using namespace std;
+
+/**
+ * Function used to write back the preproccessed data to a file.
+ *
+ * @param outputFilename Name of the file to save the data into.
+ * @param dataset The actual dataset which was read from file.
+ * @param columnDelimiter Delimiter used to separate columns of the file.
+ * @param dimensions Dimesnion which has non numeric or string data.
+ * @param encodedResult Collection of arma matrices having encoded results.
+ */
+static void WriteOutput(const string& outputFilename,
+                        const vector<vector<string>>& dataset,
+                        const string& columnDelimiter,
+                        const unordered_set<size_t>& dimensions,
+                        const unordered_map<size_t, arma::mat>& encodedResult)
```

Comment:
One of the things that makes this binding complex is that we are not actually performing string encoding for just *one* dataset, but instead we are treating every line of the input file as its own dataset.  This means that we don't have just one `arma::mat`, but instead many `arma::mat`s, all stored in the `unordered_map<size_t, arma::mat>`.  I think that it would be much simpler if we only supported encoding *one* dataset at a time.  The file format we are using is roughly like this, if I understand right:

```
1, 3, 5, Hello this is a string, 7
2, 3, 4, Hello another string which is the second dataset, 9
3, 5, 6, A third string is a third dataset, 5
...
```

i.e., we support one column (or multiple columns? I don't understand the use case for that) in an input file that are strings.  Each string is encoded as its own entire dataset.  In some sense, we are basically saying that our utility here works on *sets* of datasets.  But I think it's possible to work only on *one* dataset, if we require the following format:

```
3, 5, 7, Hello this is a string, 9
4, 5, 6, Hello a second string is the second *column* of the dataset not a *second* dataset, 8
1, 3, 5, A third string, 10
```

and in this case our output would be an mlpack matrix; for instance if we used dictionary encoding, it might be like

```
3, 5, 7, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 9
4, 5, 6, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 8
1, 3, 5, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 10
```

(I just made up all the numbers, it's not correct dictionary encoding.  But hopefully you get the idea.)

In that example, our output is just *one* dataset, with three points and many dimensions.  If a user wanted to turn a single string into an entire dataset, they could pre-tokenize it:

```
3, hello, 9
4, this, 8
5, is, 8.5
6, a, 10
2, string, 1.3
```
(Or something like that.)

And if they wanted to convert multiple strings, then they could call the utility multiple times.

The advantage of this approach is that it means we don't need any custom output writing code (we can just use `data::Save()` or even `PARAM_MATRIX_OUT` for the encoded matrix), and I think it will make the binding code itself much simpler.

Reply:
I get your point. It's like the UNIX way: "Write programs that do one thing and do it well.". I do like the idea:) In that case I have to write a utility for tokenization e.g. `preprocess_tokenize`.